qiskit/pulse/schedule.py

```diff
@@ -1504,6 +1925,33 @@ def _get_timeslots(schedule: "ScheduleComponent") -> TimeSlots:
     return timeslots
 
 
+def _get_references(block_elms: List["BlockComponent"]) -> Set[Reference]:
+    """Recursively get reference instructions in the current scope.
+
+    Args:
+        block_elms: List of schedule block elements to investigate.
+
+    Returns:
+        A set of unique reference instructions.
+    """
+    references = []
+    for elm in block_elms:
+        if isinstance(elm, ScheduleBlock):
+            references += _get_references(elm._blocks)
+        elif isinstance(elm, Reference):
+            references.append(elm)
+    return set(references)
```

Comment:
Seems odd not just to build up the set immediately?

Reply:
Good catch, definitely this is right way. Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/7a19c58b09d7e3306295b7c2da85b89569429adc